reframe/frontend/executors/policies.py

```diff
@@ -221,337 +235,404 @@ def exit(self):
+                jobs.append(t.check.build_job)
+            elif t.policy_stage == 'running':
+                jobs.append(t.check.job)
+
+        self.local_scheduler.poll(*jobs)
+
+    def advance_all(self, tasks, timeout=None):
+        t_init = time.time()
+        num_prog = 0
+
+        # progress might remove the tasks that retire or fail
+        for t in list(tasks):
+            method = getattr(self, f'advance_{t.policy_stage}')
+            num_prog += method(t)
+            t_elapsed = time.time() - t_init
+            if timeout and t_elapsed > timeout and num_prog:
+                break
 
-    def _setup_task(self, task):
+    def advance_wait(self, task):
         if self.deps_skipped(task):
             try:
                 raise SkipTestError('skipped due to skipped dependencies')
             except SkipTestError as e:
                 task.skip()
-                return False
+                self._current_tasks.remove(task)
+                return 1
+
         elif self.deps_succeeded(task):
             try:
                 task.setup(task.testcase.partition,
                            task.testcase.environ,
                            sched_flex_alloc_nodes=self.sched_flex_alloc_nodes,
                            sched_options=self.sched_options)
             except TaskExit:
-                return False
+                self._current_tasks.remove(task)
+                return 1
+
+            if isinstance(task.check, RunOnlyRegressionTest):
+                try:
+                    task.compile()
+                    task.compile_wait()
+                except TaskExit:
+                    # Run and run_wait are no-ops for
+                    # CompileOnlyRegressionTest. This shouldn't fail.
+                    self._current_tasks.remove(task)
+                    return 1
+
+                task.policy_stage = 'ready_to_run'
             else:
-                return True
+                task.policy_stage = 'ready_to_compile'
+
+            return 1
+
         elif self.deps_failed(task):
             exc = TaskDependencyError('dependencies failed')
             task.fail((type(exc), exc, None))
-            return False
+            self._current_tasks.remove(task)
+            return 1
         else:
             # Not all dependencies have finished yet
-            return False
+            return 0
 
-    def runcase(self, case):
-        super().runcase(case)
-        check, partition, environ = case
-        self._partitions.add(partition)
+    def advance_ready_to_compile(self, task):
+        if task.check.local or task.check.build_locally:
+            if len(self._local_scheduler_tasks) < self._rfm_max_jobs:
+                try:
+                    task.compile()
+                    task.policy_stage = 'compiling'
+                    self._local_scheduler_tasks.add(task)
+                except TaskExit:
+                    self._current_tasks.remove(task)
```

Comment:
The following pattern comes up again and again, so you should consider it wrapping it in a function:

```python
try:
    task.stage()
except TaskExit:
   self._current_tasks.remove(task)
```

What about perhaps something like `self._exec_stage(task.compile)`?